### DISCUSSION TITLE
How can we use elastalert to read metrics from statsd/telegraf and capture CPU and memory utlization in the alerts

### DISCUSSION THREAD

**Discussion by gyancoupa** [jertel/elastalert2#1183] [Q&A] [answered]
I saw in elastalert doc, config like statsd_host in config.yml but not much information on how to do it.

**Comment by jertel** [comment, marked as answer]
The statsd config is used for pushing performance metrics out _to_ a statsd host, not for collecting data _from_ a statsd host. ElastAlert currently only alerts on data read from an Elasticsearch or OpenSearch host.

**Reply by gyancoupa** [reply]
Is it possbile to add a piece of code inruby in elastalert rule where i can write some logic like convert timestamp found on elk logs to epoc timestamp format and then add that epoc time in the alert sent

**Reply by jertel** [reply]
You can't add custom logic in the rule itself, but you can add tiny Python [enhancements](https://elastalert2.readthedocs.io/en/latest/recipes/adding_enhancements.html#enhancements) to do this.

**Reply by gyancoupa** [reply]
cool let me try that

**Reply by gyancoupa** [reply]
I tried to have a look into this but could not figure out how will it solve my use case
I am trying to add a dynamic browser URL in my elastalert having date in epoch timestamp but before that I need to convert the date format I get from elk logs and convert it into epoc timestamp and then put it in elastalert rules file.
I know how to read from elk and read various fields from it.

**Reply by jertel** [reply]
The enhancement would format the timestamp into a special format you need within your alert message, and save it into a field such as `match["new_time"] = my_custom_formatted_timestamp`

Then in your alert body (defined in your rule yaml file) you would reference the `new_time` field.

**Reply by gyancoupa** [reply]
Ok probably I got some idea now, but where to keep this enhancements.py.
I tried to search this file on my server and found it on
/elastalert/lib/python3.7/site-packages/elastalert/enhancements.py
Do I need to make changes to this file?
Or I can create this file just in the same location where my rules.yml files are also stored and then just refer the file iinside the rules.yml file.
Can you please just put light on this part?
My rules.yml files are in different folder so where exactly to keep this enhancements.py so that I could refer it in my rules file.

**Reply by jertel** [reply]
The documentation I pasted earlier walks you through the creation of an example enhancement. Please follow those instructions. Add a `print('ENHANCING')` line to your my_enhancement.py and then run Elastalert 2 when a test rule that you know will cause an alert. Verify that you can see ENHANCING printed to your console. If you get that far then your enhancement is working,and you can work on formatting your date and placing it into a new `match['my_new_date']` field. Then reference that new field from within your rule.

**Reply by gyancoupa** [reply]
Thanks a lot, @jertel for your help and guidance, I think I am very close now, just one problem at the last stage.
So I need to add start and end date in grafana link in my elastalert based alerts, now I am able to read the end date from the epoch timestamp value converted by the code in the enhancements.py using {3} placeholder but how about the start date?
If I understand correctly at one time I can read only one value being set in the match['@timestamp'] in the enhancement.py, how to read the other timestamp value for the start date 

Pleass find below the rules.yml file
`alert_text: "*Message*: {0}\n*Grafana_Dashboard_Link*: https://*********&from=1685086200000&to={3}"
alert_text_args: ["message","stage","deployment","@timestamp"]`

The enhancement.py

`class MyEnhancement(BaseEnhancement):

    # The enhancement is run against every match
    # The match is passed to the process function where it can be modified in any way
    # ElastAlert 2 will do this for each enhancement linked to a rule
    def process(self, match):
        print('ENHANCING')
        t=match['@timestamp']
        import calendar
        from dateutil.parser import parse as date_parse
        dt =  date_parse(t)
        timestamp1 = calendar.timegm(dt.timetuple())
        epoc=timestamp1
        print("epoc is:",epoc)
        match['@timestamp']=epoc`

**Reply by jertel** [reply]
If you're trying to show a Grafana time range that includes your alert timestamp, you might want to consider creating a start/end date range that equally straddles the timestamp (it will be directly in the middle of your time range. For example, suppose your @timestamp lands on 15:45:35. You could use a 30 minute time range, and subtract half of that range (15 minutes) from the front to get your start value (15:30:35), and then add 15 minutes to the end to get your end value (16:00:35). Code example:
```
def process(self, match):
    print('ENHANCING')
    t=match['@timestamp']
    import calendar
    from dateutil.parser import parse as date_parse
    dt =  date_parse(t)
    timestamp1 = calendar.timegm(dt.timetuple())
    epoc=timestamp1
    print("epoc is:",epoc)
    
    # 30 minutes = 1800 seconds
    time_range = 1800 
    half_range = time_range/ 2
    match['start_time']= epoc - half_range
    match['end_time']= epoc + half_range
```

**Reply by gyancoupa** [reply]
Hi @jertel I have actually now got the whole logic working now using test alert method and I got to now submit the PR, now just one silly issue I am getting.

my rules.yaml  and the enhancements.py I got to keep as per the below folder structure in elastalert server and I can't keep the py file same as yaml.

elastalert/clp/rules.yaml
elastalert/enhancements/date_to_epoch

Now how can I reference the date_to_epoch.py file in a.yaml as it is present outside its parent direct?
The below syntax in rules.yaml does not work and I do not want to give an absolute path

match_enhancements:
- "../enhancements.date_to_epoch.EpochTimeConverter"

**Reply by jertel** [reply]
The enhancement directory should be a subdir of the current working directory from which Elastalert is launched. If you are using a Docker container, then the subdir should be in `/opt/elastalert/enhancements`. With that you specify the directory name 
```
match_enhancements:
  - enhancements.date_to_epoch.EpochTimeConverter
```

If you need to do something unique for your use case, where the enhancements subdir cannot exist in that location, you will need to adjust the Python sys path yourself.
